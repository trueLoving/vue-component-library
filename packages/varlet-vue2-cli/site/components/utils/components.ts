import Vue, { ComponentOptions } from 'vue'
import { CombinedVueInstance } from 'vue/types/vue'

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export interface MountComponentApi {
  instance: CombinedVueInstance<any, any, any, any, any>
  unmount(): void
}

export function mountComponent(component: ComponentOptions<Vue>, container = 'body', options = {}): MountComponentApi {
  const instance = new (Vue.extend(component))(options)
  const el = instance.$mount().$el
  const wrapper = document.querySelector(container)!
  wrapper.appendChild(el)

  return {
    instance,
    unmount() {
      instance.$destroy()
      wrapper.removeChild(el)
    },
  }
}
