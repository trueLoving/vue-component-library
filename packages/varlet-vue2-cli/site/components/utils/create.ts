import Vue, { ComponentOptions, RenderContext, VueConstructor } from 'vue'
import { SlotsMixin } from './mixins/slots'
import { ListenersMixin } from './mixins/listeners'

export type ComponentOptionsPlugin = ComponentOptions<Vue> & { install?: (app: VueConstructor) => void }

export function nextTick() {
  return new Promise(Vue.nextTick)
}

export function mergeSlots(context: RenderContext) {
  const scopedSlots = context.scopedSlots ?? context.data.scopedSlots ?? {}
  const slots = context.slots()

  Object.keys(slots).forEach((key) => {
    if (!scopedSlots[key]) {
      scopedSlots[key] = () => slots[key]
    }
  })

  return scopedSlots
}

export function defineComponent(sfc: ComponentOptionsPlugin & { name: string }) {
  if (sfc.mixins) {
    sfc.mixins.push(SlotsMixin)
    sfc.mixins.push(ListenersMixin)
  } else {
    sfc.mixins = [SlotsMixin, ListenersMixin]
  }

  return sfc
}
