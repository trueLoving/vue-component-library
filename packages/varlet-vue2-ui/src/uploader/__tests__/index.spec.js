import example from '../example'
import Uploader from '..'
import VarUploader from '../Uploader'
import ImagePreview from '../../image-preview'
import { mount } from '@vue/test-utils'
import Vue from 'vue'
import { delay, mockFileReader, mockStubs } from '../../utils/jest'

const createEvent = (filename) => {
  return {
    target: {
      files: [new File([], filename)],
    },
  }
}

test('test uploader example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
})

test('test uploader plugin', () => {
  Vue.use(Uploader)
  expect(Vue.component(Uploader.name)).toBeTruthy()
})

test('test uploader onAfterRead', async () => {
  const onAfterRead = jest.fn()

  const wrapper = mount(VarUploader, {
    listeners: {
      afterRead: onAfterRead,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onAfterRead).toHaveBeenCalledTimes(1)

  wrapper.destroy()
})

test('test uploader onBeforeRead', async () => {
  const onAfterRead = jest.fn()

  const wrapper = mount(VarUploader, {
    listeners: {
      afterRead: onAfterRead,
      beforeRead: (file) => file.name.endsWith('jpg'),
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onAfterRead).toHaveBeenCalledTimes(0)

  await wrapper.vm.handleChange(createEvent('cat.jpg'))
  expect(onAfterRead).toHaveBeenCalledTimes(1)

  wrapper.destroy()
})

test('test uploader preview', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const { mockRestore: mockRestoreStubs } = mockStubs()
  const onInput = jest.fn((value) => wrapper.setProps({ value }))

  const wrapper = mount(VarUploader, {
    propsData: {
      value: [],
    },
    listeners: {
      input: onInput,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.jpg'))
  await delay(16)
  await wrapper.find('.var-uploader__file').trigger('click')
  await delay(100)

  expect(document.querySelector('.var-popup').style.display).toBe('')
  ImagePreview.close()
  await delay(300)

  mockRestoreStubs()
  wrapper.destroy()
  mockRestore()
})

test('test uploader onOversize', async () => {
  const onOversize = jest.fn()

  const wrapper = mount(VarUploader, {
    propsData: {
      maxsize: -1,
    },
    listeners: {
      oversize: onOversize,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onOversize).toHaveBeenCalledTimes(1)

  wrapper.destroy()
})

test('test uploader onRemove', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onInput = jest.fn((value) => wrapper.setProps({ value }))
  const onRemove = jest.fn()

  const wrapper = mount(VarUploader, {
    propsData: {
      value: [],
    },
    listeners: {
      remove: onRemove,
      input: onInput,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onInput).toHaveBeenCalledTimes(1)

  await wrapper.find('.var-uploader__file-close').trigger('click')
  expect(onInput).toHaveBeenCalledTimes(2)
  expect(onRemove).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.value).toStrictEqual([])

  wrapper.destroy()
  mockRestore()
})

test('test uploader onBeforeRemove', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onInput = jest.fn((value) => wrapper.setProps({ value }))
  const onBeforeRemove = jest.fn(() => false)
  const onRemove = jest.fn()

  const wrapper = mount(VarUploader, {
    propsData: {
      value: [],
    },
    listeners: {
      beforeRemove: onBeforeRemove,
      remove: onRemove,
      input: onInput,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onInput).toHaveBeenCalledTimes(1)

  await wrapper.find('.var-uploader__file-close').trigger('click')
  expect(onBeforeRemove).toHaveBeenCalledTimes(1)
  expect(onInput).toHaveBeenCalledTimes(1)
  expect(onRemove).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.value).toHaveLength(1)

  wrapper.destroy()
  mockRestore()
})

test('test uploader validation', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onInput = jest.fn((value) => wrapper.setProps({ value }))

  const wrapper = mount(VarUploader, {
    propsData: {
      value: [],
      rules: [(v) => v.length >= 1 || '您至少上传一个'],
    },
    listeners: {
      input: onInput,
    },
  })

  wrapper.vm.validate()
  await delay(16)
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find('.var-form-details__message').text()).toBe('您至少上传一个')

  await wrapper.vm.handleChange(createEvent('cat.png'))
  await delay(16)
  expect(onInput).toHaveBeenCalledTimes(1)
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find('.var-form-details__message').exists()).toBeFalsy()

  wrapper.vm.reset()
  await delay(16)
  expect(wrapper.vm.value).toStrictEqual([])

  wrapper.destroy()
  mockRestore()
})

test('test uploader disabled', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onInput = jest.fn((value) => wrapper.setProps({ value }))
  const onRemove = jest.fn()
  const onAfterRead = jest.fn()

  const wrapper = mount(VarUploader, {
    propsData: {
      disabled: false,
      value: [],
    },
    listeners: {
      afterRead: onAfterRead,
      remove: onRemove,
      input: onInput,
    },
  })

  expect(wrapper.html()).toMatchSnapshot()
  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onAfterRead).toHaveBeenCalledTimes(1)

  await wrapper.setProps({ disabled: true })
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find('.var-uploader__action-input[disabled]').exists()).toBeTruthy()

  await wrapper.find('.var-uploader__file-close').trigger('click')
  expect(onRemove).toHaveBeenCalledTimes(0)

  wrapper.destroy()
  mockRestore()
})

test('test uploader length over maxlength in multiple mode', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onInput = jest.fn((value) => wrapper.setProps({ value }))

  const wrapper = mount(VarUploader, {
    propsData: {
      maxlength: 1,
      value: [],
    },
    listeners: {
      input: onInput,
    },
  })

  expect(wrapper.html()).toMatchSnapshot()

  const event = {
    target: {
      files: [new File([], 'cat.png'), new File([], 'dog.png')],
    },
  }

  await wrapper.vm.handleChange(event)
  expect(wrapper.vm.value.length).toBe(1)

  wrapper.destroy()
  mockRestore()
})
