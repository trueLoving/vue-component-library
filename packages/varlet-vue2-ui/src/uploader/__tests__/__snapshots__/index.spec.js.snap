// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test uploader disabled 1`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <!---->
  </transition-stub>
  <transition-stub name=\\"var-fade\\">
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub name=\\"var-pop-center\\"></transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test uploader disabled 2`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">
        cat.png
      </div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-uploader__file-close-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
      <!---->
      <div class=\\"var-uploader__file-indicator\\"></div>
    </div>
    <div class=\\"var--relative var-uploader__action var-elevation--2 var-uploader--disabled\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\" disabled=\\"disabled\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <!---->
  </transition-stub>
  <transition-stub name=\\"var-fade\\">
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub name=\\"var-pop-center\\"></transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test uploader example 1`] = `
"<div class=\\"var-uploader-example\\">
  <div class=\\"app-type\\">基本使用</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">文件预览</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var-uploader__file var-elevation--2\\">
        <div class=\\"var-uploader__file-name\\">
          https://varlet-vue2.vercel.app/cat.jpg
        </div>
        <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-uploader__file-close-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div> <img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" class=\\"var-uploader__file-cover\\">
        <div class=\\"var-uploader__file-indicator\\"></div>
      </div>
      <div class=\\"var-uploader__file var-elevation--2\\">
        <div class=\\"var-uploader__file-name\\">
          https://www.runoob.com/try/demo_source/mov_bbb.mp4
        </div>
        <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-uploader__file-close-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div> <img src=\\"https://varlet-vue2.vercel.app/cover.jpg\\" class=\\"var-uploader__file-cover\\">
        <div class=\\"var-uploader__file-indicator\\"></div>
      </div>
      <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">上传状态</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var-uploader__file var-elevation--2 var-uploader--loading\\">
        <div class=\\"var-uploader__file-name\\">
          https://varlet-vue2.vercel.app/cat.jpg
        </div>
        <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-uploader__file-close-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div> <img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" class=\\"var-uploader__file-cover\\">
        <div class=\\"var-uploader__file-indicator\\"></div>
      </div>
      <div class=\\"var-uploader__file var-elevation--2\\">
        <div class=\\"var-uploader__file-name\\">
          https://varlet-vue2.vercel.app/cat.jpg
        </div>
        <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-uploader__file-close-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div> <img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" class=\\"var-uploader__file-cover\\">
        <div class=\\"var-uploader__file-indicator var-uploader--success\\"></div>
      </div>
      <div class=\\"var-uploader__file var-elevation--2\\">
        <div class=\\"var-uploader__file-name\\">
          https://varlet-vue2.vercel.app/cat.jpg
        </div>
        <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-uploader__file-close-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div> <img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" class=\\"var-uploader__file-cover\\">
        <div class=\\"var-uploader__file-indicator var-uploader--error\\"></div>
      </div>
      <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">文件数量限制</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <div class=\\"var-form-details\\">
        <div class=\\"var-form-details__message\\"></div>
        <div class=\\"var-form-details__length\\">0 / 1</div>
      </div>
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">文件大小限制</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">上传预处理</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">禁用</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var--relative var-uploader__action var-elevation--2 var-uploader--disabled\\"><input type=\\"file\\" accept=\\"image/*\\" disabled=\\"disabled\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">只读</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" disabled=\\"disabled\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">删除前处理</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var-uploader__file var-elevation--2\\">
        <div class=\\"var-uploader__file-name\\">
          https://varlet-vue2.vercel.app/cat.jpg
        </div>
        <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-uploader__file-close-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div> <img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" class=\\"var-uploader__file-cover\\">
        <div class=\\"var-uploader__file-indicator\\"></div>
      </div>
      <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">自定义上传样式</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var--relative\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <button class=\\"var-button var--box var-button--normal var--inline-flex var-button--primary var-elevation--2\\">
          <!---->
          <div class=\\"var-button__content\\">上传</div>
        </button></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type\\">字段校验</div>
  <div class=\\"var-uploader var--box\\">
    <div class=\\"var-uploader__file-list\\">
      <div class=\\"var-uploader__file var-elevation--2\\">
        <div class=\\"var-uploader__file-name\\">
          https://varlet-vue2.vercel.app/cat.jpg
        </div>
        <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-uploader__file-close-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div> <img src=\\"https://varlet-vue2.vercel.app/cat.jpg\\" class=\\"var-uploader__file-cover\\">
        <div class=\\"var-uploader__file-indicator var-uploader--error\\"></div>
      </div>
      <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
    <transition-stub name=\\"var-fade\\">
      <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
        <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
        <transition-stub name=\\"var-pop-center\\"></transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"space\\"></div>
</div>"
`;

exports[`test uploader length over maxlength in multiple mode 1`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\"></div>
      <div class=\\"var-form-details__length\\">0 / 1</div>
    </div>
  </transition-stub>
  <transition-stub name=\\"var-fade\\">
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub name=\\"var-pop-center\\"></transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test uploader validation 1`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\">您至少上传一个</div>
      <div class=\\"var-form-details__length\\"></div>
    </div>
  </transition-stub>
  <transition-stub name=\\"var-fade\\">
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub name=\\"var-pop-center\\"></transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test uploader validation 2`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">
        cat.png
      </div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-uploader__file-close-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
      <!---->
      <div class=\\"var-uploader__file-indicator\\"></div>
    </div>
    <div class=\\"var--relative var-uploader__action var-elevation--2\\"><input type=\\"file\\" accept=\\"image/*\\" class=\\"var-uploader__action-input\\"> <i class=\\"var-icon var-uploader__action-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <!---->
  </transition-stub>
  <transition-stub name=\\"var-fade\\">
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub name=\\"var-pop-center\\"></transition-stub>
    </div>
  </transition-stub>
</div>"
`;
